### PR TITLE
add nft_metadata_update event to NEP171

### PR DESCRIPTION
**Pull request by lachlanglen**
See NEAR Forum discussion here: https://gov.near.org/t/extending-nep-171-events-standard-to-include-nft-update-events/30285

- Add `nft_metadata_update` event to NEP-171 events
- Increment events version to `1.2.0`

---

## NEP Status *(Updated by NEP moderators)*

SME reviews:

* [x] @telezhnaya https://github.com/near/NEPs/pull/469#pullrequestreview-1348563704
* [x] @frol https://github.com/near/NEPs/pull/469#issuecomment-1476817621

Contract Standards WG voting indications:

* 👍 @frol https://github.com/near/NEPs/pull/469#issuecomment-1476817621
* 👍 @mfornet https://github.com/near/NEPs/pull/469#pullrequestreview-1350224466
* 👍 @abacabadabacaba https://github.com/near/NEPs/pull/469#issuecomment-1504232783
* 👍 @fadeevab
* 👍 @robert-zaremba  https://github.com/near/NEPs/pull/469#pullrequestreview-1396040781

### CODE REVIEW

**Comment by frol** [comment]
Thank you @lachlanglen for submitting this NEP extension.

As a moderator, I reviewed this NEP and it meets the proposed template guidelines. I am moving this NEP to the REVIEW stage and would like to ask the @near/wg-contract-standards working group members to assign 2 Technical Reviewers to complete a technical review (see expectations). If you want to assign yourself, please mention that you are acting as the Technical Reviewers.

Please find some guidelines below for completing the technical review.

<details>
<summary>Technical Review Guidelines</summary>


* First, review the proposal within one week. If you have any suggestions that could be fixed, leave them as comments to the author. It may take a couple of iterations to resolve any open comments. 

* Second, once all the suggestions are addressed, produce a Technical Summary, which helps the working group members make a weighted decision faster. Without the summary, the working group will have to read the whole discussion and potentially miss some details.

Technical Summary guidelines:

  * A recommendation for the working group if the NEP is ready for voting (it could be approving or rejecting recommendation). Please note that this is the reviewer's personal recommendation.

  * A summary of benefits that surfaced in previous discussions. This should include a concise list of all the benefits that others raised, not just the ones that the reviewer personally agrees with.

  *  A summary of concerns or blockers, along with their current status and resolution. Again, this should reflect the collective view of all commenters, not just the reviewer's perspective.

Here is a [nice example](https://github.com/near/NEPs/pull/399#issuecomment-1462492128) and a template for your convenience:

```

## Example

### Recommendation

Add recommendation

### Benefits

* Benefit

* Benefit

### Concerns

| # | Concern | Resolution | Status |

| - | - | - | - |   

| 1 | Concern | Resolution | Status |

| 2 | Concern | Resolution | Status |

```

Please tag the @near/nep-moderators once you are done, so we can move this NEP to the voting stage. Thanks again.

</details>

File: neps/nep-0171.md
Comment (telezhnaya) [inline neps/nep-0171.md:256]: Have we considered the option to include the updated NFT in the logs?
At the first glance, I see obvious pros - no need to make the request to see updated NFT, and obvious cons - the log is growing, while we are trying to save the letters. 
I don't have the answer here, just want to know your opinion.

File: neps/nep-0171.md
Comment (telezhnaya) [inline neps/nep-0171.md:221]: A token is updated if the `owner_id` is changed, because `owner_id` is one of the Token fields: https://nomicon.io/Standards/Tokens/NonFungibleToken/Metadata. 
So, for the transfer, we have to emit 2 logs, which is weird.

Do we want to use `token_metadata_update`?

File: neps/nep-0171.md
Comment (frol) [reply neps/nep-0171.md:256]: @telezhnaya It was discussed on the gov forum (see the link in the PR description). The major concern I see with posting more details into the logs is that the media might not fit into the logs. I think we are inevitably heading into the direction where we will use the logs to quickly catch if a relevant change happened, but then to learn more details you would need to execute some view function (as an optimization, we could even consider following the state updates on the indexer side, and execute view functions locally)

**Comment by frol** [comment]
As a working group member, I’d like to nominate @telezhnaya and myself (@frol) as the Subject Matter Experts to review this NEP.

File: neps/nep-0171.md
Comment (telezhnaya) [reply neps/nep-0171.md:256]: Ah, I see. I totally forgot about the media, you are right.

File: neps/nep-0171.md
Comment (lachlanglen) [reply neps/nep-0171.md:221]: This is a very good point. What about `nft_metadata_update` to keep the naming convention for NFT-related events?

Comment (lachlanglen) [reply neps/nep-0171.md:221]: Just updated to `nft_metadata_update`

**Review by telezhnaya — APPROVED** [review]

**Comment by frol** [comment]
### Recommendation

As a subject matter expert, I recommend approving this NEP as it is a non-breaking quality-of-life improvement.

### Benefits

* Apps that cache indexed NFTs will benefit from having this new event type as they won't need to have a custom logic to track potentially changing NFTs or refetch all NFTs on every transaction to NFT contract
* This NEP extension only introduces an additional event type, so there is no breaking change to the original NEP

### Concerns

| # | Concern | Resolution | Status |
| - | - | - | - |   
| 1 | Ecosystem will be split where legacy contracts won't emit these new events, so legacy support will still be needed | In the future, there will be fewer legacy contracts and eventually apps will have support for this type of event | Resolved |

**Review by mfornet — COMMENTED** [review]
As a WG member, I lean towards approving this NEP.

File: neps/nep-0171.md
Comment (robert-zaremba) [reply neps/nep-0171.md:221]: Yes, let's do normal formatting. We describe an action here, rather than an object.

File: neps/nep-0171.md
Comment (robert-zaremba) [inline neps/nep-0171.md:218]: Shouldn't we have a changelog?

File: neps/nep-0171.md
Comment (robert-zaremba) [inline neps/nep-0171.md:234]: Side comment for the future events (not the NEP-171, as it is already set).
Personally, I don't think we need to have `nft_` prefix in all the event names (not the struct names... they are not relevant to the standard). The NFT namespace is already implicity with the `standard: "nep171"`.

File: neps/nep-0171.md
Comment (robert-zaremba) [inline neps/nep-0171.md:279]:
```suggestion
// An event emitted when token metadata is updated. 
```

Comment (robert-zaremba) [reply neps/nep-0171.md:279]: I don't think _log_ is appropriate here. This is an object, not a log.

**Review by robert-zaremba — APPROVED** [review]
I agree with the update and the version presented here:
* metadata should not be part of the event (noting here about it because there was a discussion about it). 

In general, I always advice to put the minimum amount of data to identify what the event is tracking. Everything else can be fetched from the transaction data. Let's remember that events are designed for off chain systems (eg indexers)

File: neps/nep-0171.md
Comment (frol) [reply neps/nep-0171.md:218]: Do you refer to the NEP changelog? There is a placeholder there already. We will fill it out once we have all the reviews, so we can compile all the feedback and the final decision details into it.

File: neps/nep-0171.md
Comment (lachlanglen) [reply neps/nep-0171.md:234]: I disagree, I think this prefix makes it easier for indexers to handle events by type, and it helps categorize events within the contract.

File: neps/nep-0171.md
Comment (lachlanglen) [reply neps/nep-0171.md:279]: I don't think it's a big deal, probably easier to just keep it consistent with existing comments. I personally don't find it confusing.

Comment (robert-zaremba) [reply neps/nep-0171.md:279]: logging and event != emitting an event. In blockchian context we talk about emitting an event usually. So if we want to be consistent with that, then we should use _emit_

File: neps/nep-0171.md
Comment (robert-zaremba) [reply neps/nep-0171.md:218]: Yes. wouldn't it be better to have it already set and be part of the final review?

Comment (frol) [reply neps/nep-0171.md:218]: @robert-zaremba Ideally, yes, but I cannot see how to make it productive using GitHub UI. Feel free to review [my SME review comment](https://github.com/near/NEPs/pull/469#issuecomment-1476817621).

File: neps/nep-0171.md
Comment (frol) [reply neps/nep-0171.md:279]: @robert-zaremba Let's create a separate PR with this improvement to the terminology, so we can update all the Events NEPs at once and discuss it separately from this NEP.

**Comment by lachlanglen** [comment]
Hi @frol - any update on next steps for this?

**Review by fadeevab — APPROVED** [review]

**Comment by abacabadabacaba** [comment]
As a working group member, I lean towards approving this NEP.

**Comment by ori-near** [comment]
As the moderator, I would like to thank the @lachlanglen for submitting this Contract Standards Standards NEP extension,  and for the @near/wg-contract-standards working group members for your review. Based on the voting indications above, it seems like this proposal is close to a decision. We are therefore scheduling a Contract Standards Work Group call this week, where this NEP can enter the final voting stage. Anyone can discuss the technical details by adding your comments to this discussion. And join the call to learn about the final decision and how to get more involved.

**Meeting Info**
📅 Date: Friday, April 21, 3:15pm UTC
✏️ [Register](https://bit.ly/41dNPmM)

File: neps/nep-0171.md
Comment (robert-zaremba) [inline neps/nep-0171.md:285]: in my opinion this should not be needed - this is already in the transaction, and I don't think memo is something that should be indexed. However since this is in other NEP-171 events, we can keep it here.

**Review by robert-zaremba — APPROVED** [review]
Approving. Clarified the concern about `memo` field in events.

File: neps/nep-0171.md
Comment (frol) [reply neps/nep-0171.md:285]: `memo` is not a necessary part of the transaction. There is even no function associated with this event as there is no meaningful one.

**Comment by frol** [comment]
🎉 This NEP was approved by @near/wg-contract-standards members. I will wait until we have a recording published to update the NEP document with the references and merge this PR.

**Review by frol — APPROVED** [review]

**Comment by frol** [comment]
Thank you to everyone who attended the fourth Contract Standards Working Group meeting last Friday! The working group members reviewed the NEP and reached the following consensus:

**Status:** Approved

* 👍 @frol https://github.com/near/NEPs/pull/469#issuecomment-1476817621
* 👍 @mfornet https://github.com/near/NEPs/pull/469#pullrequestreview-1350224466
* 👍 @abacabadabacaba https://github.com/near/NEPs/pull/469#issuecomment-1504232783
* 👍 @fadeevab
* 👍 @robert-zaremba  https://github.com/near/NEPs/pull/469#pullrequestreview-1396040781

Meeting Recording:
https://youtu.be/KOIT8XDQNjM

@lachlanglen Thank you for authoring this NEP!

Next Steps:
- [x] NEP moderator to embed the benefits & concerns into the NEP document
- [ ] @near/nep-moderators To approve and merge the NEP

**Comment by nearbuild** [comment]
@frol can this move into Approved NEPs

**Pull request merged**
